csrc/preseg_passes/move_pad.cpp

```diff
@@ -0,0 +1,530 @@
+  if (pad->out()->uses().size() == 1 && pad->out()->uses()[0]->isA<CatOp>()) {
+    return true;
+  }
+  for (Val* pad_val : pad->getPadWidths()) {
+    if (!simplifyExpr(
+             SimplifyingIrBuilder::geExpr(pad_val, pad->fusion()->zeroVal()))
+             ->isTrue()) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// check if `use` and `p` are the same pad operatoin.
+bool isSamePadOp(Expr* use, PadOp* p) {
+  if (!use->isA<PadOp>()) {
+    return false;
+  }
+
+  auto* use_pad = use->as<PadOp>();
+  if (!simplifyExpr(SimplifyingIrBuilder::eqExpr(use_pad->value(), p->value()))
+           ->isTrue()) {
+    return false;
+  }
+
+  if (use_pad->getPadWidths().size() != p->getPadWidths().size()) {
+    return false;
+  }
+
+  std::vector<int64_t> padded_axes = p->getPaddedAxes();
+  if (use_pad->getPaddedAxes() != padded_axes) {
+    return false;
+  }
+
+  for (auto idx : padded_axes) {
+    if (!simplifyExpr(
+             SimplifyingIrBuilder::eqExpr(
+                 use_pad->getPadWidths(idx).first, p->getPadWidths(idx).first))
+             ->isTrue() ||
+        !simplifyExpr(SimplifyingIrBuilder::eqExpr(
+                          use_pad->getPadWidths(idx).second,
+                          p->getPadWidths(idx).second))
+             ->isTrue()) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+// This operation replaces:
+//   CatOp(inputs)
+// with:
+//   BinaryOp(inputs[n-1], BinaryOp(inputs[n-2], BinaryOp(..., BinaryOp[0])...))
+Val* replayCatOpWithBinaryOp(
+    const std::vector<Val*>& inputs,
+    DataType data_type) {
+  // replay `CatOp` with series of BinaryOp instead, since we might have
+  // pushed `PadOp` out and breaking the codegen if `CatOp` remains.
+  Val* res = nullptr;
+  bool is_boolean = isBooleanType(data_type);
+  for (Val* inp : inputs) {
+    if (res == nullptr) {
+      res = inp;
+    } else {
+      if (is_boolean) {
+        res = bitwise_or(res, inp);
+      } else {
+        res = add(res, inp);
+      }
+    }
+  }
+  // restore data type if it's promoted by BinaryOp.
+  return maybeCastOp(data_type, res);
+}
+
+// The pass assumes propagating PadOp with zero pad. The criteria here for
+// return true is that `unaryOp(0) == 0`
+bool zeroIsFixedPoint(UnaryOpType t) {
+  switch (t) {
+    case UnaryOpType::Cast:
+    case UnaryOpType::Abs:
+    case UnaryOpType::Asin:
+    case UnaryOpType::Asinh:
+    case UnaryOpType::Atan:
+    case UnaryOpType::Atanh:
+    case UnaryOpType::Ceil:
+    case UnaryOpType::Erf:
+    case UnaryOpType::Erfinv:
+    case UnaryOpType::Floor:
+    case UnaryOpType::Gelu:
+    case UnaryOpType::Imag:
+    case UnaryOpType::Silu:
+    case UnaryOpType::Neg:
+    case UnaryOpType::Real:
+    case UnaryOpType::Relu:
+    case UnaryOpType::Round:
+    case UnaryOpType::Sin:
+    case UnaryOpType::Sinh:
+    case UnaryOpType::Sqrt:
+    case UnaryOpType::Tan:
+    case UnaryOpType::Tanh:
+    case UnaryOpType::Trunc:
```

Comment:
Add `Expm1`, `Log1p`, `Frac`, `Signbit`, and reorder `Silu` to preserve alphabetic order.
```suggestion
    case UnaryOpType::Abs:
    case UnaryOpType::Asin:
    case UnaryOpType::Asinh:
    case UnaryOpType::Atan:
    case UnaryOpType::Atanh:
    case UnaryOpType::Ceil:
    case UnaryOpType::Erf:
    case UnaryOpType::Erfinv:
    case UnaryOpType::Expm1:
    case UnaryOpType::Floor:
    case UnaryOpType::Frac:
    case UnaryOpType::Gelu:
    case UnaryOpType::Imag:
    case UnaryOpType::Log1p:
    case UnaryOpType::Neg:
    case UnaryOpType::Real:
    case UnaryOpType::Relu:
    case UnaryOpType::Round:
    case UnaryOpType::Silu:
    case UnaryOpType::Signbit:
    case UnaryOpType::Sin:
    case UnaryOpType::Sinh:
    case UnaryOpType::Sqrt:
    case UnaryOpType::Tan:
    case UnaryOpType::Tanh:
    case UnaryOpType::Trunc:
```
Not sure what you'd like to do with `IsInf`, `IsNegInf`, `IsPosInf`, and `IsNan`. These take numeric 0 and give bool `false` so I believe it's safe to include them here.